### ISSUE TITLE
Partial construction status

### ISSUE THREAD

**Issue by jim-huynh** [CleverRaven/Cataclysm-DDA#26211]
I was thinking about the current state of the faction camp, and how constructing a faction camp currently is very inflexible - your kitchen will always look the same.

What is the feasibility of adding a blueprint status to buildings, and completion states in ticks rather than separate construction states that are all or nothing?

If the completion status is in ticks, NPCs could hypothetically eventually build other parts of the base while you build a separate construction item, or while you're crafting but still in the reality bubble.

In this system you could interrupt construction, go take the NPCs out hunting, and then come back and let them complete stuff, or even have another NPC complete the work.

I have not started implementing this, and this is just a discussion to see whether or not it would be a useful addition.

**Comment by mlangsdorf** [comment]
See here: https://github.com/mlangsdorf/Cataclysm-DDA/issues/18

tl;dr:
Yes, it's a good idea.  I would be happy to collaborate with you if you want to work on this.

**Comment by jim-huynh** [comment]
Ok I'll look into it.  Your proposal is a little different than what I'm suggesting though.

To make things simple, resources are consumed in the construction immediately, rather than bits and pieces.  This is to get around the math problem of when resources are consumed.

Also, the task of getting NPCs to build constructions is outside of the scope.

**Comment by mlangsdorf** [comment]
Here's the logic for consuming resources piecemeal:
* You want to build a stone house, and set an NPC builder to work at it. House has 12 walls, 4 doors, and 9 roofs (it's a 5x5 square house with a door in the middle of each wall).  It takes 48 logs for the walls, 20 two-by-fours + 80 nails for the doors, and another 18 logs and 36 heavy sticks for the roof.
* NPC starts work on it, building the NW corner and wall to the east of that, then skips the door (it's not legal to build it yet) and builds the wall on the east side of the north door.
* You decide to train some fabrication and install the door yourself.
* The NPC continues working and completes the NE corner, just in time for a hulk to show up and punch you through it. The NPC patiently rebuilds the NE corner and without interruption, completes the house.

So then how many logs did it take the NPC to build the house?  If it's piecemeal, that's an easy answer. If they're all consumed at once, he consumed less logs than he should have and more two-by-fours and nails.

**Comment by mlangsdorf** [comment]
But you don't like my proposal. Which is fine, I have lots of bad ideas.

My current state of the code is at #26296.  That brings the code up to "not awful."

My next steps:
* Fix logging so the NPC turns 100% of the cut down trees into logs, and takes 100% of the logs back to base.  Fairly straightforward.
* Add foraging, trapping, and hunting tasks.  Again, fairly straightforward except I decided to clean up the code first.
* Break the dependency on the camp overseer, so he can be replaced by a bulletin board. This is 90% complete: in the current code, the camp manager is an access point to talk_function::companion_mission() and a storage system for companion_mission_role_id and the camp's omt_pos3.   Those can be accessed in other ways with a little code tweaking, except the camp manager also stores the sort points.  My plan is to use the neglected base_camp code, which already stores an omt_pos3, as the new storage space for camp data.
* Stop using mapgen to upgrade camps. I'm pretty sure acidia used mapgen to upgrade camps so he'd have a unique ID to determine the camp upgrade level, plus he was a big mapgen guy.  I think that was a bad decision and want to replace it with an integer value stored in the expanded base_camp structure.

Please tell me how we can work together on this, or why my idea is a bad idea, or failing that, what's the best way to work on the same set of code without stomping on each other's toes.

**Comment by jim-huynh** [comment]
=/  I said that it's different, not bad or any other adjective.  I thought that having a construction consume the entirety of the resources needed at the start will prevent the need to display which resources are missing if a resource decides to go missing at some time between the start and the end of construction.  It is a lazy implementation, but I thought it would be simpler.  

I suppose the more intuitive solution would be piece-meal consumption.  

Ideally, your work and my work would only overlap in very minor ways.  I'm adding a blueprint status to the map for the player (or npc) to activate and build over time.  Your NPC improvements would allow them to hopefully activate existing blueprints to complete it.  I don't intend on modifying existing faction building plans.

A blueprint would work somewhat like how the current kiln/smoker system works.  You set up some structure (as a blueprint, it is immediate), and then activate it to construct it.  It exists on the map before completion, which allows you to work on it intermittently.

A question that remains is what happens when a hulk comes by and destroys it.  Does it return 75% of resources consumed?  50%?

**Comment by kevingranade** [comment]
> A question that remains is what happens when a hulk comes by and destroys it. Does it return 75% of resources consumed? 50%?

Probably something between pretend the wall was finished and then smash it, or smash some of the components.  I.e. invoke some existing destruction code.

**Comment by jim-huynh** [comment]
> 
> 
> > A question that remains is what happens when a hulk comes by and destroys it. Does it return 75% of resources consumed? 50%?
> 
> Probably something between pretend the wall was finished and then smash it, or smash some of the components. I.e. invoke some existing destruction code.

Invoking any existing destruction code sounds like the correct way to do it.

**Comment by mlangsdorf** [comment]
Okay, I think misunderstood your proposal, but with a second pass, I have a better grasp.
So you activate a blueprint for the 5x5 building mentioned above and get something like:
```
##x##
#   #
x   x
#   #
##x##
```
and then you can 'e'xamine any particular tile to get that's part of the construction blueprint for a 5x5 house and has some construction requirements, which if you fulfill you can then choose to construct. So you might end up with:
```
r-+-n
|   #
x   x
|   #
##x##
```
as the player builds the west and north walls, but discovers he ran out of nails before installing the west door.

It's definitely a good idea, and I'm going to want blueprints eventually for my NPC builders.

I still think piecemeal resources are the way to go - you may have to rebuild the same wall multiple times, or you may not want to build exactly to the blueprint for some reason or another.

**Comment by kevingranade** [comment]
We have a system for incrementally building structures via blueprints now.
